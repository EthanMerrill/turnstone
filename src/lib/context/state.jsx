import React, { createContext, useReducer, useEffect } from 'react'
import reducer from '../reducers/reducer'
import {setQuery} from '../actions/actions'
import undef from '../utils/undef'

const StateContext = createContext()

const StateContextProvider = (props) => {
  const { styles = {}, text = '', items = [] } = props
  const { children, ...propsMinusChildren} = props
  const [state, dispatch] = useReducer(reducer, {
    query: text,
    items,
    itemsError: false,
    canShowListbox: false,
    highlighted: undef,
    selected: undef,
    customStyles: styles,
    props: propsMinusChildren
  })

  useEffect(() => dispatch(setQuery(text)), [text]) // TODO: Is this needed?

  return (
    <StateContext.Provider
      value={{state, dispatch}}>
      {children}
    </StateContext.Provider>
  )
}

export { StateContext, StateContextProvider }
