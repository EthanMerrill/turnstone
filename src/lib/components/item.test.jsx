import React from 'react'
import renderer from 'react-test-renderer'

import { describe, expect, test } from 'vitest'
import { StateContextProvider } from '../context/state.jsx'
import Item from './item.jsx'

// TODO: Create test helper files?
const customStyles = {
  item: 'item-class',
  highlightedItem: 'highlighted-item-class',
  topItem: 'top-item-class',
  split: 'split-class',
  splitChar: 'split-char-class'
}
const item = {
  text: 'Chicago, Illinois, United States',
  value: {
    name: 'Chicago, Illinois, United States',
    coords: '41.882304590139135, -87.62327214400634'
  }
}

const index = 0

const component = renderer.create(
  <StateContextProvider styles={customStyles} text={'Chi'} items={[item]}>
   <Item
      index={index}
      key={`item${index}`}
      item={item}
    />
  </StateContextProvider>
)
let tree = component.toJSON()

describe('Item', () => {
  test('Component renders correctly', () => {
    expect(tree).toMatchSnapshot()
  })

  test('Link changes the class when hovered', () => {
    // Manually trigger onMouseEnter and rerender
    tree.props.onMouseEnter()
    tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})