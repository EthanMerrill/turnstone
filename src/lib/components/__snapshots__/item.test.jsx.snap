// Vitest Snapshot v1

exports[`Item > Component renders correctly 1`] = `
<div
  aria-label="Chicago, Illinois, United States"
  className="item-class top-item-class"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  role="option"
  style={
    {
      "overflow": "hidden",
      "whiteSpace": "nowrap",
    }
  }
>
  Chicago, Illinois, United States
</div>
`;

exports[`Item > Link changes the class when hovered 1`] = `
<div
  aria-label="Chicago, Illinois, United States"
  aria-selected={true}
  className="highlighted-item-class top-item-class"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  role="option"
  style={
    {
      "overflow": "hidden",
      "whiteSpace": "nowrap",
    }
  }
>
  Chicago, Illinois, United States
</div>
`;
